internal/storage/inmem/query.go

```diff
@@ -0,0 +1,187 @@
+package inmem
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/consul/internal/storage"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+)
+
+// indexSeparator delimits the segments of our radix tree keys.
+const indexSeparator = "\x00"
+
+// readIndexer implements the memdb.Indexer interface (see FromArgs).
+type readIndexer struct{}
+
+// FromArgs constructs a radix tree key from an ID for lookup.
+func (i readIndexer) FromArgs(args ...any) ([]byte, error) {
+	if l := len(args); l != 1 {
+		return nil, fmt.Errorf("expected 1 arg, got: %d", l)
+	}
+	id, ok := args[0].(*pbresource.ID)
+	if !ok {
+		return nil, fmt.Errorf("expected *pbresource.ID, got: %T", args[0])
+	}
+	return indexFromID(id), nil
+}
+
+// idIndexer implements the memdb.SingleIndexer and memdb.PrefixIndexer
+// interfaces. It is used for indexing resources by their IDs.
+type idIndexer struct{ readIndexer }
+
+// FromObject constructs a radix tree key from a Resource at write-time, or an
+// ID at delete-time.
+func (i idIndexer) FromObject(raw any) (bool, []byte, error) {
+	switch t := raw.(type) {
+	case *pbresource.ID:
+		return true, indexFromID(t), nil
+	case *pbresource.Resource:
+		return true, indexFromID(t.Id), nil
+	}
+	return false, nil, fmt.Errorf("expected *pbresource.Resource or *pbresource.ID, got: %T", raw)
+}
+
+// PrefixFromArgs constructs a radix tree key prefix from a query for listing.
+func (i idIndexer) PrefixFromArgs(args ...any) ([]byte, error) {
+	if l := len(args); l != 1 {
+		return nil, fmt.Errorf("expected 1 arg, got: %d", l)
+	}
+
+	q, ok := args[0].(query)
+	if !ok {
+		return nil, fmt.Errorf("expected query, got: %T", args[0])
+	}
+	return q.indexPrefix(), nil
+}
+
+// ownerIndexer implements the memdb.SingleIndexer interface. It is used for
+// indexing resources by their owners.
+type ownerIndexer struct{ readIndexer }
+
+// FromObject constructs a radix key tree from a Resource at write-time.
+func (i ownerIndexer) FromObject(raw any) (bool, []byte, error) {
+	res, ok := raw.(*pbresource.Resource)
+	if !ok {
+		return false, nil, fmt.Errorf("expected *pbresource.Resource, got: %T", raw)
+	}
+	if res.Owner == nil {
+		return false, nil, nil
+	}
+	return true, indexFromID(res.Owner), nil
+}
+
+func indexFromType(t storage.UnversionedType) []byte {
+	var b indexBuilder
+	b.String(t.Group)
+	b.String(t.Kind)
+	return b.Bytes()
+}
+
+func indexFromTenancy(t *pbresource.Tenancy) []byte {
+	var b indexBuilder
+	b.String(t.Partition)
+	b.String(t.PeerName)
+	b.String(t.Namespace)
```

Comment:
To avoid having to implement this twice (and handle the ensuing OSS→ENT merge hell) I was thinking of keeping the partition and namespace plumbing in OSS, but having the brains/controllers live in ENT only, and rejecting non-default values at the RPC handle level.

Does that sound ok, @mkeeler?

Reply:
starting a slack convo...

Reply:
Keeping this as-is for the moment 🧑‍⚖️ 

I really like @mkeeler's idea of having separate stores (or even backends) for each partition, and think this is ultimately the way to go, but it comes with some complexity that I'd like to defer for now so we can make progress on the other parts of this project.

As separate stores makes cross-partition watches impossible, or at least more difficult, need to either run a controller per-partition or have controllers watch resources in multiple partitions, and then reconfigure all of this on the fly as partitions are created and deleted.